src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.BasicBlockBuilder.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis
+{
+    internal partial class ControlFlowGraphBuilder
+    {
+        internal sealed class BasicBlockBuilder
+        {
+            private readonly ArrayBuilder<IOperation> _statementsBuilder;
+            private readonly HashSet<BasicBlockBuilder> _predecessorsBuilder;
+            public (IOperation Condition, bool JumpIfTrue, Branch Branch) InternalConditional;
+            public (IOperation Value, Branch Branch) InternalNext;
+            
+            public BasicBlockBuilder(BasicBlockKind kind)
+            {
+                Kind = kind;
+                _statementsBuilder = ArrayBuilder<IOperation>.GetInstance();
+                _predecessorsBuilder = new HashSet<BasicBlockBuilder>();
+            }
+
+            public IReadOnlyCollection<IOperation> Statements => _statementsBuilder;
+            public ImmutableArray<IOperation> GetImmutableStatements() => _statementsBuilder.ToImmutableAndFree();
+            public ISet<BasicBlockBuilder> Predecessors => _predecessorsBuilder;
+            public BasicBlockKind Kind { get; }
+
+            public int Ordinal { get; set; } = -1;
+
+            public bool IsReachable { get; set; } = false;
+
+            public ControlFlowRegion Region { get; set; }
+
+            internal void AddStatement(IOperation statement)
+            {
+                _statementsBuilder.Add(statement);
+            }
+
+            internal void AddStatements(IEnumerable<IOperation> statements)
```

Comment:
@AlekseyTs were you planning on adjusting this in another followup?

Reply:
Ah, I see that you are.

---
In reply to: [192534549](https://github.com/dotnet/roslyn/pull/27315#discussion_r192534549) [](ancestors = 192534549)